pandas/core/generic.py

```diff
@@ -1995,10 +1996,13 @@ def empty(self) -> bool_t:
     def __array__(self, dtype: npt.DTypeLike | None = None) -> np.ndarray:
         values = self._values
         arr = np.asarray(values, dtype=dtype)
-        if arr is values and using_copy_on_write():
-            # TODO(CoW) also properly handle extension dtypes
-            arr = arr.view()
-            arr.flags.writeable = False
+        if arr is values and using_copy_on_write() and self._mgr.is_single_block:
+            # Check if self._values coerced data
+            if not is_1d_only_ea_dtype(self.dtypes.iloc[0]) or not is_numeric_dtype(
+                self.dtypes.iloc[0]
```

Comment:
We have 2 different cases here:

- NumPy dtypes that are caught by the is_1d_only_ea_dtype
- EA dtypes -> they get coerced to object by self._values above, so we only have to catch cases where a coercion to object does not trigger a copy.

astype_is_view would work when we get rid of the conversion to object in the middle. I'll clarify the comment though.

Reply:
So for the EA dtypes, you rely on the assumption that numeric dtypes were cast to object. But in theory, someone could implement its own EA which is indicated to be "numeric" but does not do this conversion to object dtype.

Now, in general I think we are lacking a part of the story around having proper information about copies/views with generic EAs (that was also the case when doing the `astype_is_view`)

Reply:
That's actually what our numeric arrow dtypes do, but since those rely on pyarrow to convert itself to numpy arrays, those are already set to readonly in case they are a view:

```
In [91]: arr = pd.array([1, 2], dtype=pd.ArrowDtype(pa.int64()))

In [92]: np_arr = np.asarray(arr)

In [93]: np_arr
Out[93]: array([1, 2])

In [94]: np_arr[0] = 100
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [94], in <cell line: 1>()
----> 1 np_arr[0] = 100

ValueError: assignment destination is read-only
```

Reply:
Fair point, there is actual a more elegant solution. We can check if both steps can be done without copying via astype_is_view.

Reply:
Yeah I ran into this as well (that they set themselves to read only)